pkg/controller/data_upload_controller.go

```diff
@@ -0,0 +1,498 @@
+		return []reconcile.Request{}
+	}
+
+	requests := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: du.Namespace,
+			Name:      du.Name,
+		},
+	}
+	return []reconcile.Request{requests}
+}
+
+func (r *DataUploadReconciler) patchDataUpload(ctx context.Context, req *velerov2alpha1api.DataUpload, mutate func(*velerov2alpha1api.DataUpload)) error {
+	original := req.DeepCopy()
+	mutate(req)
+	if err := r.client.Patch(ctx, req, client.MergeFrom(original)); err != nil {
+		return errors.Wrap(err, "error patching DataUpload")
+	}
+
+	return nil
+}
+
+func prepareDataUpload(du *velerov2alpha1api.DataUpload) {
+	du.Status.Phase = velerov2alpha1api.DataUploadPhasePrepared
+}
+
+func (r *DataUploadReconciler) errorOut(ctx context.Context, du *velerov2alpha1api.DataUpload, err error, msg string, log logrus.FieldLogger) (ctrl.Result, error) {
+	if se, ok := r.snapshotExposerList[du.Spec.SnapshotType]; ok {
+		se.CleanUp(ctx, getOwnerObject(du), du.Spec.CSISnapshot.VolumeSnapshot, du.Spec.SourceNamespace)
+	} else {
+		err = errors.Wrapf(err, "failed to clean up exposed snapshot with could not find %s snapshot exposer", du.Spec.SnapshotType)
+	}
+	return ctrl.Result{}, r.updateStatusToFailed(ctx, du, err, msg, log)
+}
+
+func (r *DataUploadReconciler) updateStatusToFailed(ctx context.Context, du *velerov2alpha1api.DataUpload, err error, msg string, log logrus.FieldLogger) error {
+	original := du.DeepCopy()
+	du.Status.Phase = velerov2alpha1api.DataUploadPhaseFailed
+	du.Status.Message = errors.WithMessage(err, msg).Error()
+	if du.Status.StartTimestamp.IsZero() {
+		du.Status.StartTimestamp = &metav1.Time{Time: r.clock.Now()}
+	}
+
+	du.Status.CompletionTimestamp = &metav1.Time{Time: r.clock.Now()}
+	if err = r.client.Patch(ctx, du, client.MergeFrom(original)); err != nil {
+		log.WithError(err).Error("error updating DataUpload status")
+		return err
+	}
+
+	return nil
+}
+
+func (r *DataUploadReconciler) acceptDataUpload(ctx context.Context, du *velerov2alpha1api.DataUpload) (bool, error) {
+	updated := du.DeepCopy()
+	updated.Status.Phase = velerov2alpha1api.DataUploadPhaseAccepted
+
+	r.logger.Infof("Accepting snapshot backup %s", du.Name)
+
+	time.Sleep(2 * time.Second)
+
+	err := r.client.Update(ctx, updated)
+	if err == nil {
+		r.logger.WithField("Dataupload", du.Name).Infof("This datauplod backup has been accepted by %s", r.nodeName)
+		return true, nil
+	} else if apierrors.IsConflict(err) {
+		r.logger.WithField("Dataupload", du.Name).Info("This datauplod backup has been accepted by others")
+		return false, nil
+	} else {
+		return false, err
+	}
+}
+
+func (r *DataUploadReconciler) closeDataPath(ctx context.Context, duName string) {
+	fsBackup := r.dataPathMgr.GetAsyncBR(duName)
+	if fsBackup != nil {
+		fsBackup.Close(ctx)
+	}
+
+	r.dataPathMgr.RemoveAsyncBR(duName)
+}
+
+func (r *DataUploadReconciler) setupExposeParam(du *velerov2alpha1api.DataUpload) interface{} {
```

Comment:
No need to put this function as the method of struct `DataUploadReconciler `.

Reply:
For Expose interface `Expose(context.Context, corev1.ObjectReference, string, time.Duration, interface{}) error` which is an interface and it's unrelated to one specific type of exposer, but it needs to pass interface{} type of value which is related to one specific type of exposer.  

which is awkward, so we need one function(setupExposeParam) to prepare the value of the interface{}. Here we have the `CSI Snapshot exposer`,  maybe later we'll have other kinds of exposer. in order to organize unseen values of the interface{} for other types of exposer in the future, I need to pass all values into setupExposeParam.

Reply:
If the parameters used by other types of exposers in the future are accessible from the `DataUploadReconciler` structure, then I think that's fair enough to keep it as the structure's method.